doc/README.md

```diff
@@ -184,12 +184,64 @@ In that case, please open an issue about the particular documentation convention
   }
   ```
 
-- Use [definition lists](#definition-lists) to document function arguments, and the attributes of such arguments. For example:
+- When showing inputs/outputs of any [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop), such as a shell or the Nix REPL, use a format as you'd see in the REPL, while trying to visually separate inputs from outputs.
+  This means that for a shell, you should use a format like the following:
+  ```shell
+  $ nix-build -A hello '<nixpkgs>' \
+    --option require-sigs false \
+    --option trusted-substituters file:///tmp/hello-cache \
+    --option substituters file:///tmp/hello-cache
+  /nix/store/zhl06z4lrfrkw5rp0hnjjfrgsclzvxpm-hello-2.12.1
+  ```
+  Note how the input is preceded by `$` on the first line and indented on subsequent lines, and how the output is provided as you'd see on the shell.
+
+  For the Nix REPL, you should use a format like the following:
+  ```shell
+  nix-repl> builtins.attrNames { a = 1; b = 2; }
+  [ "a" "b" ]
+  ```
+  Note how the input is preceded by `nix-repl>` and the output is provided as you'd see on the Nix REPL.
+
+- When documenting functions or anything that has inputs/outputs and example usage, use nested headings to clearly separate inputs, outputs, and examples.
+  Keep examples as the last nested heading, and link to the examples wherever applicable in the documentation.
+  An example:
+  ```
+  ## buildImage
+
+  Some explanation about the function here.
+  Describe a particular scenario, and point to [](#ex-dockerTools-buildImage), which is an example demonstrating it.
+
+  ### Inputs
+
+  Documentation for the inputs of `buildImage`.
+  Perhaps even point to [](#ex-dockerTools-buildImage) again when talking about something specifically linked to it.
+
+  ### Passthru outputs
+
+  Documentation for any passthru outputs of `buildImage`.
+
+  ### Examples
+
+  Note that this is the last nested heading in the `buildImage` section.
+
+  :::{.example #ex-dockerTools-buildImage}
+
+  # Using `buildImage`
+
+  Example of how to use `buildImage` goes here.
+
+  :::
+  ```
+
+- Use [definition lists](#definition-lists) to document function arguments, the attributes of such arguments, or any other attribute sets in general. For example:
 
   ```markdown
   # pkgs.coolFunction
 
   Description of what `coolFunction` does.
+
+  ## Inputs
```

Comment:
Regarding function inputs/outputs. I am not so sure about ditching `f :: A -> b -> C` style notation.

Some reasons why it's nice (to me):

* it's very consistent in style, which is helpful when trying to quickly find info
* it's easy to auto-generate (given some doc strings conventions as current RFCs are pushing, or possibly, eventually: a type system)
* it helps communicate that it is very possible to partially evaluate functions. So for instance, if `repeatN :: String -> Integer -> String`, and `"hello" :: String`, we have `repeatN "hello" :: Integer -> String`. Put differently, there are three, equally valid ways to look at `repeatN`: 
1. `repeatN` is a function which takes `String` as input and produces another function `Integer -> String` as output, or 
2. `repeatN` is a function which takes a function `String -> Integer` as input and produces a `String` as output, or 
3. `repeatN` is a function which takes a `String` and `Integer` as input and produces a `String` as output
* it is already a convention used by some Nix IDE tools and documentation
* it is already a convention used by some languages (e.g. Haskell), and similar-ish to the [BNF](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form) convention more generally 

Some reasons why its not nice (to me):

* it takes some getting used to (but its knowledge that is helpful in other languages too)
* nothing is said immediately about what `String` is required (having some way to include variable names along with the types would be helpful...)

But I can imagine a convention where we write along with `runCommand :: String -> AttrSet -> String -> Derivation` something like `runCommand :: name -> drvAttr -> buildCommand -> Derivation` so that we have overall:

```
runCommand :: name -> drvAttr -> command -> Derivation
                        :: String -> AttrSet -> String -> Derivation

Returns a derivation realizing a binary executing specified shell commands. 

name :: String
    The name that Nix will append to the store path in the same way that stdenv.mkDerivation uses its name attribute.

drvAttr :: AttrSet
    Attributes to pass to the underlying call to stdenv.mkDerivation.

buildCommand :: String
    Shell commands to run in the derivation builder.
    
Note
    You have to create a file or directory $out for Nix to be able to run the builder successfully.
```

Reply:
"ditching" is a strong word - there hasn't been any convention before this, so there's nothing to "ditch".
In any case, I think this comment misunderstands what the convention is about. This is something to provide a common structure to make it easier for people to navigate any part of the manual. It doesn't say how to write the content of each heading.

There is another convention to use definition lists for arguments and attributes, which is also there to provide a common structure for readers. Applied to your example, the `name`, `drvAttr`, and `buildCommand` inputs should ideally be structured in a definition list. That still doesn't force anyone to "ditch" the style notation you like.

Reply:
I updated the PR with another clarification to avoid this kind of misunderstanding again.

Reply:
I do not think your changes address the question I raised. In the interest of clarity, note that:

* no statements were made regarding an already existing convention. There are a few different "local conventions". A reference was being made to [*one* of these local conventions](https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-runCommand). For clarity, I'll refer to these "local conventions" as "dialects".

* no statements were made regarding personal preference for an existing documentation dialect over another. An attempt was made to express uncertainty around the selection of one dialect over others, without due diligence. 

* (this will be fixed) no statements were made regarding collating a list of existing documentation-dialects, or better yet, to link to existing research on this front.

To fix the last point:
* The `nixos-render-doc` work *might* have had such a collation (not easy to (re)find, therefore, not included for now). 
* [`nixdoc`](https://github.com/nix-community/nixdoc), which is already used for documenting items in [`lib`, generates output like so](https://nixos.org/manual/nixpkgs/unstable/#sec-functions-library-strings)

To fix potential issues around misunderstanding of colloquialisms, note that "to ditch" means:
> To [discard](https://en.wiktionary.org/wiki/discard) or [abandon](https://en.wiktionary.org/wiki/abandon).

I will stick to discard from now on for clarity.

The convention here---even though a work in progress---has the potential to influence new PRs suggesting changing existing `f :: A -> B -> C` structure to use a Input/Output header structure. This is one sense in which `f :: A -> B -> C` style structure may be discarded.

If a convention is ratified by this PR, then it would greenlight changing existing documentation differing from the approved structure. This is another sense in which existing dialects could be discarded.

Therefore, two issues are being raised: 

1. Please do not select a convention, without a summary of existing dialects, and some analysis of their pros and cons. 

2. Please mark clearly that this document is a work in progress that does not endorse any convention yet.